src/GafferScene/Set.cpp

```diff
@@ -192,75 +194,105 @@ IECore::ConstInternedStringVectorDataPtr Set::computeSetNames( const Gaffer::Con
 {
 	ConstInternedStringVectorDataPtr inNamesData = inPlug()->setNamesPlug()->getValue();
 
-	InternedString name = namePlug()->getValue();
-	if( !name.string().size() )
+	const std::string &names = namePlug()->getValue();
+	if( !names.size() )
 	{
 		return inNamesData;
 	}
 
-	const std::vector<InternedString> &inNames = inNamesData->readable();
-	if( std::find( inNames.begin(), inNames.end(), name ) != inNames.end() )
+	if( modePlug()->getValue() == Remove )
 	{
 		return inNamesData;
 	}
 
-	if( modePlug()->getValue() == Remove )
+	InternedStringVectorDataPtr result = NULL;
+
+	Tokenizer namesTokenizer( names, boost::char_separator<char>( " " ) );
+	const std::vector<InternedString> &inNames = inNamesData->readable();
+	std::vector<InternedString>::const_iterator inBegin = inNames.begin();
+	std::vector<InternedString>::const_iterator inEnd = inNames.end();
+	for( Tokenizer::iterator name = namesTokenizer.begin(), end = namesTokenizer.end(); name != end; name++ )
+	{
+		if( std::find( inBegin, inEnd, *name ) != inNames.end() )
+		{
+			continue;
+		}
+
+		if ( result == NULL )
+		{
+			// we are modifying the list, make a copy now
+			result = inNamesData->copy();
+		}
+
+		result->writable().push_back( *name );
+	}
+
+	if ( result == NULL )
 	{
+		// output is the same as the input
 		return inNamesData;
 	}
 
-	InternedStringVectorDataPtr result = inNamesData->copy();
-	result->writable().push_back( name );
 	return result;
 }
 
 void Set::hashSet( const IECore::InternedString &setName, const Gaffer::Context *context, const ScenePlug *parent, IECore::MurmurHash &h ) const
 {
-	if( setName != namePlug()->getValue() )
+	Tokenizer namesTokenizer( namePlug()->getValue(), boost::char_separator<char>( " " ) );
```

Comment:
I'd suggest avoiding boost::tokenizer again here. Actually, since `hashSet()` is more performance critical than any of the others, I wonder if we should avoid tokenizing entirely (because it needs to allocate memory for the vector, if not the InternedStrings), and instead search for the token directly? If we don't do that right away, let's at least use `tokenize( names, vector<InternendString> )` followed by `find()`, and add a todo with the idea.